javaslang/src/test/java/javaslang/collection/IteratorTest.java

```diff
@@ -360,6 +360,11 @@ public void shouldGenerateInfiniteStreamBasedOnSupplierWithAccessToPreviousValue
         assertThat(iterate(2, (i) -> i + 2).take(3).reduce((i, j) -> i + j)).isEqualTo(12);
     }
 
+    @Test
+    public void shouldNotCallSupplierUntilNecessary() {
+        assertThat(iterate(2, (i) -> {throw new RuntimeException("The supplier function should not be called since the next value is not needed in order to get head.");}).head()).isEqualTo(2);
```

Comment:
my preference would be to ommit the text in the Exception, the name of the test already explains the intent :)